Sources/Host/Linux/X86_64/PTraceX86_64.cpp

```diff
@@ -107,12 +109,24 @@ state32_to_user(user_fpregs_struct &user,
   //
   // SSE State
   //
-  user.mxcsr = state.sse.mxcsr;
-  user.mxcr_mask = state.sse.mxcsrmask;
-  auto xmm_space = reinterpret_cast<uint8_t *>(user.xmm_space);
+  xfpregs.fpregs.mxcsr = state.sse.mxcsr;
+  xfpregs.fpregs.mxcr_mask = state.sse.mxcsrmask;
+  auto xmm_space = reinterpret_cast<uint8_t *>(xfpregs.fpregs.xmm_space);
   for (size_t n = 0; n < array_sizeof(state.sse.regs); n++) {
     memcpy(xmm_space + n * 16, &state.sse.regs[n], sizeof(state.sse.regs[n]));
   }
+
+  //
+  //  EAVX State
+  //
+  auto ymmh = reinterpret_cast<uint8_t *>(xfpregs.ymmh);
+  for (size_t n = 0; n < array_sizeof(state.avx.regs); n++) {
+    size_t sseSize = sizeof(state.sse.regs[n]);
+    size_t avxSize = sizeof(state.avx.regs[n]);
+    auto avxHigh =
+        reinterpret_cast<const uint8_t *>(&state.avx.regs[n]) + sseSize;
+    memcpy(ymmh + n * 16, avxHigh, avxSize - sseSize);
```

Comment:
Yeah that's kind of what I meant with "What is this?" :)

I recommend using `static const` in these. e.g. `static const size_t
sseSize = sizeof(state.sse.regs[n]);`. Same for `avxSize`. And then you
could have one more `static const` variable for the size of the top half of
the AVX register.

On Wed, May 18, 2016 at 10:22 AM, Francis Ricci <notifications@github.com>
wrote:

> In Sources/Host/Linux/X86_64/PTraceX86_64.cpp
> <https://github.com/facebook/ds2/pull/290#discussion_r63745197>:
>
> >    for (size_t n = 0; n < array_sizeof(state.sse.regs); n++) {
> >      memcpy(xmm_space + n * 16, &state.sse.regs[n], sizeof(state.sse.regs[n]));
> >    }
> > +
> > +  //
> > +  //  EAVX State
> > +  //
> > +  auto ymmh = reinterpret_cast<uint8_t *>(xfpregs.ymmh);
> > +  for (size_t n = 0; n < array_sizeof(state.avx.regs); n++) {
> > +    size_t sseSize = sizeof(state.sse.regs[n]);
> > +    size_t avxSize = sizeof(state.avx.regs[n]);
> > +    auto avxHigh =
> > +        reinterpret_cast<const uint8_t *>(&state.avx.regs[n]) + sseSize;
> > +    memcpy(ymmh + n * 16, avxHigh, avxSize - sseSize);
>
> the size of one ymmh vector. I assume I could actually just use avxSize -
> sseSize again there.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/ds2/pull/290/files/4cdd106d838cdce1c9def04b399a00dad6f8c852..e30974e02b13f69cecb0e60929796545d77f0752#r63745197>
>



-- 
Stephane Sezer